hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/TestReplicationManagerScenarios.java

```diff
@@ -0,0 +1,783 @@
+import org.apache.hadoop.ozone.protocol.commands.ReplicateContainerCommand;
+import org.apache.hadoop.ozone.protocol.commands.SCMCommand;
+import org.apache.ozone.test.TestClock;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_SCM_WAIT_TIME_AFTER_SAFE_MODE_EXIT;
+import static org.mockito.ArgumentMatchers.any;
+
+/**
+ * This class tests the replication manager using a set of scenarios defined in
+ * JSON files. The scenarios define a container and a set of replicas, and the
+ * expected results from the replication manager. The scenarios are defined in
+ * JSON files in the test resources directory. The test files are loaded in
+ * {@link #init()} and the tests are run in {@link #testAllScenarios(Scenario)}.
+ *
+ * There are several inner class defined within this class, and they are used to
+ * deserialize the JSON files into Java objects. In general any field which is a
+ * setter on the inner class can be set in the JSON file.
+ *
+ * TODO - The framework does not allow for testing mis-replicated containers.
+ */
+
+public class TestReplicationManagerScenarios {
+  private static final Map<String, UUID> ORIGINS = new HashMap<>();
+  private static final Map<String, DatanodeDetails> DATANODE_ALIASES
+      = new HashMap<>();
+  private static final Map<DatanodeDetails, NodeStatus> NODE_STATUS_MAP
+      = new HashMap<>();
+  private static final String TEST_RESOURCE_PATH = "/replicationManagerTests";
+  private static final List<Scenario> TEST_SCENARIOS = new ArrayList<>();
+
+  private Map<ContainerID, Set<ContainerReplica>> containerReplicaMap;
+  private Set<ContainerInfo> containerInfoSet;
+  private ContainerReplicaPendingOps containerReplicaPendingOps;
+  private Set<Pair<UUID, SCMCommand<?>>> commandsSent;
+
+  private OzoneConfiguration configuration;
+  private ReplicationManager replicationManager;
+  private LegacyReplicationManager legacyReplicationManager;
+  private ContainerManager containerManager;
+  private PlacementPolicy ratisPlacementPolicy;
+  private PlacementPolicy ecPlacementPolicy;
+  private EventPublisher eventPublisher;
+  private SCMContext scmContext;
+  private NodeManager nodeManager;
+  private TestClock clock;
+  private ReplicationManagerReport repReport;
+  private ReplicationQueue repQueue;
```

Comment:
Can we make these local variables? They don't have any special setup requiring extra code.

Reply:
Yea, I guess they can be locals as they are just used in the test method.